mechanics/mechanics.py

```diff
@@ -56,7 +60,16 @@ def resolve_battle(board1: Board, board2: Board) -> Tuple[int, int, List[dict]]:
         attack = randint(0, attacking_token.unit.attack)
         defense = randint(0, defending_token.unit.defense)
 
+        distance = defending_token.distance(attacking_token)
+        attack_range = attacking_token.unit.range
+        if distance <= 0:
+            raise ValueError("Distance between tokens is 0")
+
+        damage_reduction_divider = calculate_damage_divider(distance, attack_range)
+
         damage = max(0, attack - defense)
```

Comment:
we could also set minimal damage to 1 and implement additional logic for dodge (e.g. with given probability dependent on magical power etc.)